XdmodAID.md

```diff
@@ -0,0 +1,868 @@
+
+
+# Basic Processing
+
+The overall processing flow is as follows
+```
+                  OpenStack                               OpenShift                 KeyCloak        ColdFront
+
+      xdmod-openstack   xdmod-openstack-hypervisors     xdmod-openshift                  xdmod-hierarchy
+
+      xdmod-shredder        xdmod-shredder              xdmod-shredder                   xdmod-import-csv
+
+
+                           xdmod-ingestor
+
+                                                       aggregate-supremm
+
+                                      acl-config
+
+                                      xdmod-ui
+
+                                batch_export_manager.php
+```
+
+
+# xdmod-openstack-hypervisors
+This script fetches the hypervisor information from openstack used in calculating
+percent usage.
+
+This was the simpliest script and required very few changes to work.
+
+There was a bug in the command line section of this script that is present in the
+upstream version, which we decided to fix with our version of the script. So our
+version will be different and possibly incompatible with the upstream version.
+
+This script produces a file specified on the command line that will need to be
+shredded.
+
+# acl-config
+This is a small utility that rebuilds the consistency between the configuration files
+and the database.  It is needed to be run as there are several processes that
+modify one or the other but not both.  It is an example of a work-a-round that
+became part of the processing.  From a conversation I had in a meeting with one of
+the xdmod developers, it is expected to eventually be eliminated, however it is
+currently low priority.
+
+As the xdmod-ui requires the database and configration files to be consistent,
+and we do not currently support RWX volumes, this needs to be run in parallel
+to the UI.
+
+Most of the time it runs withing a couple of minutes, sometimes it takes longer
+and will ocassionally fail due to a time out when it cannot lock certain tables.
+
+# xdmod-openstack
+
+The script xdmod-openstack required to be rewritten as the one that was there
+was using ceilometer.  Ceilometer collected metrics that could be used for
+billing.  With ceilometer we could have had usage based billing.  Unfortunately,
+ceilometer is no longer included in OpenStack.
+
+In order to get the information about VMs, we went with the nova API.  Since the
+Nova API only gives information on VMs that are not deleted, by collectings data
+every 5 minutes, we expected to to catch when VMs topped running/were deleted.
+over time we have had to relax the performance requirements as due to the
+non-performant filesystem.  It is currently set to run every 20 minutes, but
+even that is probably a bit optimistic.
+
+In both the "OpenShift" and the "Generic" structure there is a way to
+track mounted volumes to the VM.  From an initial glance, something like
+this should be done to track volume useage, doing so here will not track
```

Comment:
will not -> will

Reply:
The next line for context is "volumes that are allocated, but not mounted."

I guess I need to include the data structure, however, it assumes that any volume is attached.  There is no provision made for unattached volumes.  This is why implementing storage metrics is important.